File: ZerolanPlayground/Assets/Scripts/Data/Protocol.cs
using Newtonsoft.Json;

namespace Data
{
    public class ZerolanProtocol<T>
    {
        public string Protocol { get; set; }
        public string Version { get; set; }
        public string Message { get; set; }
        public string Action { get; set; }
        public int Code { get; set; }
        public T Data { get; set; }

        [JsonConstructor]
        public ZerolanProtocol(string protocol, string version, string message, string action, int code, T data)
        {
            Protocol = protocol;
            Version = version;
            Message = message;
            Action = action;
            Code = code;
            Data = data;
        }
    }
}

File: zerolan-data/src/zerolan/data/protocol/protocol.py
from typing import TypeVar

from pydantic import BaseModel, Field

T = TypeVar('T')


class ZerolanProtocol(BaseModel):
    """
    Represents a message following the Zerolan protocol.
    """
    protocol: str = Field("ZerolanProtocol", description="The name of the protocol.")
    version: str = Field("1.1", description="The version of the protocol.")
    message: str = Field(..., description="A descriptive message.")
    action: str = Field(..., description="The action associated with the message.")
    code: int = Field(..., description="A status code.")
    data: T = Field(..., description="The data payload, which can be of any type.")
